File: algorithms/python/validParentheses/validParentheses.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/valid-parentheses/
# Author: Miao Zhang
# Date:   2021-01-06

class Solution:
    def isValid(self, s: str) -> bool:
        dicts = {')': '(', '}': '{', ']': '['}
        stack = []
        for c in s:
            if c in dicts.values():
                stack.append(c)
            elif c in dicts.keys():
                if stack == [] or stack.pop() != dicts[c]:
                    return False
            else:
                return False
        return stack == []

File: algorithms/cpp/validParentheses/validParentheses.cpp
// Source: https://leetcode.com/problems/valid-parentheses/
// Author: Miao Zhang
// Date:   2021-01-06

class Solution {
public:
    bool isValid(string s) {
        map<char, char> m{{')', '('}, {'}', '{'}, {']', '['}};
        stack<char> st;
        for (auto c: s) {
            if (c == '(' || c == '{' || c == '[') {
                st.push(c);
            } else if (c == ')' || c == '}' || c == ']' ) {
                if (st.empty() || st.top() != m[c]) return false;
                st.pop();
            } else {
                return false;
            }
        }
        return st.empty();
    }
};
